core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessfulRouter.java

```diff
@@ -269,19 +273,26 @@ private CoreEvent eventWithCurrentContextDeleted(CoreEvent event) {
 
   private Function<Throwable, Throwable> getThrowableFunction(CoreEvent event) {
     return throwable -> {
-      Throwable cause = getMessagingExceptionCause(throwable);
-      CoreEvent exceptionEvent = event;
+      CoreEvent exhaustionCauseEvent = event;
       if (throwable instanceof MessagingException) {
-        exceptionEvent = ((MessagingException) throwable).getEvent();
+        exhaustionCauseEvent = ((MessagingException) throwable).getEvent();
       }
-      return new MessagingException(exceptionEvent,
-                                    new RetryPolicyExhaustedException(createStaticMessage(UNTIL_SUCCESSFUL_MSG_PREFIX,
-                                                                                          cause.getMessage()),
-                                                                      cause, owner),
-                                    owner);
+      Throwable exhaustionCause = getMessagingExceptionCause(throwable);
+      Throwable retryExhaustedException =
+          new RetryPolicyExhaustedException(createStaticMessage(UNTIL_SUCCESSFUL_MSG_PREFIX, exhaustionCause.getMessage()),
+                                            exhaustionCause, owner);
+      CoreEvent retryExhaustedEvent = getRetryExhaustedEvent(exhaustionCauseEvent, retryExhaustedException);
+      return new MessagingException(retryExhaustedEvent, retryExhaustedException, owner);
     };
   }
 
+  private CoreEvent getRetryExhaustedEvent(CoreEvent exhaustionCauseEvent, Throwable retryExhaustedException) {
+    ErrorBuilder errorBuilder = ErrorBuilder.builder(retryExhaustedException)
+        .errorType(errorTypeLocator.lookupErrorType(retryExhaustedException));
```

Comment:
have the result of `errorTypeLocator.lookupErrorType(retryExhaustedException)` already available during initialization , to avoid doing the lookup fir every handled error.